website/blog/2025-04-01-tlm-tracing/index.mdx

```diff
@@ -0,0 +1,416 @@
+
+MLflow provides tracing and evaluation capabilities that can be used to monitor, review, and debug the performance of AI applications. This post will show how to apply Cleanlab's TLM to LLM responses recorded with MLflow tracing. Using Cleanlab's TLM with MLflow enables you to systematically log, track, and analyze the trustworthiness evaluations provided by TLM for your LLM interactions.
+
+You can find a notebook version of this guide [here](https://github.com/cleanlab/cleanlab-tools/blob/main/TLM-MLflow-Integration/evaluating_traces_TLM_mlflow_dl.ipynb).
+
+:::info
+
+This guide requires a Cleanlab TLM API key. If you don't have one, you can sign up for a free trial [here](https://tlm.cleanlab.ai/).
+:::
+
+## Install dependencies & Set environment variables
+
+To work through this guide, you'll need to install the MLflow, OpenAI, and Cleanlab TLM Python packages:
+
+```bash
+pip install -q mlflow openai cleanlab-tlm --upgrade
+```
+
+Next, import the dependencies:
+
+```python
+import mlflow
+import os
+import json
+import pandas as pd
+
+from rich import print
+from openai import OpenAI
+from getpass import getpass
+```
+
+### API Keys
+
+This guide requires two API keys:
+
+- [OpenAI API Key](https://platform.openai.com/api-keys)
+- [Cleanlab TLM API Key](https://tlm.cleanlab.ai/)
+
+If they are not already set as environment variables, you can set them manually as follows:
+
+```python
+if not (openai_api_key := os.getenv("OPENAI_API_KEY")):
+    openai_api_key = getpass("🔑 Enter your OpenAI API key: ")
+if not (cleanlab_tlm_api_key := os.getenv("CLEANLAB_TLM_API_KEY")):
+    cleanlab_tlm_api_key = getpass("🔑 Enter your Cleanlab TLM API key: ")
+
+os.environ["OPENAI_API_KEY"] = openai_api_key
+os.environ["CLEANLAB_TLM_API_KEY"] = cleanlab_tlm_api_key
+```
+
+### Set Up MLflow Tracking Server and Logging
+
+To manage our experiments, parameters, and results effectively, we'll [start a local MLflow Tracking Server](https://mlflow.org/docs/latest/getting-started/logging-first-model/step1-tracking-server). This provides a dedicated UI for [monitoring and managing our experiments](https://mlflow.org/docs/latest/tracking/) and allows us to [configure MLflow to connect to this server](https://mlflow.org/docs/latest/getting-started/). We'll then enable autologging for OpenAI to automatically capture relevant information from our API calls.
+
+```python
+# This will start a server on port 8080, in the background
+# Navigate to http://localhost:8080 to see the MLflow UI
+%%bash --bg
+mlflow server --host 127.0.0.1 --port 8080
+```
+
+```python
+# Set up MLflow tracking server
+mlflow.set_tracking_uri("http://localhost:8080")
+
+# Enable logging for OpenAI SDK
+mlflow.openai.autolog()
+
+# Set experiment name
+mlflow.set_experiment("Eval OpenAI Traces w/ TLM")
```

Comment:
Can we remove the slash? This isn't permitted in some backends (filestore in Windows, for instance)